File: 351002/Pisarik/task1/app/services/message_service.py
from typing import List, Optional
from app.models.message import Message
from app.models.news import News
from app.repositories.in_memory_repository import InMemoryRepository
from app.dtos.message_request import MessageRequestTo
from app.dtos.message_response import MessageResponseTo

class MessageService:
    def __init__(self, repo: InMemoryRepository[Message], news_repo: InMemoryRepository[News]) -> None:
        self._repo = repo
        self._news_repo = news_repo

    def _ensure_news_exists(self, news_id: int) -> None:
        news = self._news_repo.get_by_id(news_id)
        if not news:
            raise ValueError(f"News with id {news_id} not found")

    def create_message(self, dto: MessageRequestTo) -> MessageResponseTo:
        self._ensure_news_exists(dto.newsId)
        entity = Message(id=self._repo.next_id(), newsId=dto.newsId, content=dto.content)
        created = self._repo.create(entity)
        return MessageResponseTo(id=created.id, newsId=created.newsId, content=created.content)

    def get_message(self, message_id: int) -> Optional[MessageResponseTo]:
        entity = self._repo.get_by_id(message_id)
        if not entity:
            return None
        return MessageResponseTo(id=entity.id, newsId=entity.newsId, content=entity.content)

    def get_all_messages(self) -> List[MessageResponseTo]:
        return [MessageResponseTo(id=i.id, newsId=i.newsId, content=i.content) for i in self._repo.get_all()]

    def get_messages_by_news(self, news_id: int) -> List[MessageResponseTo]:
        self._ensure_news_exists(news_id)
        return [MessageResponseTo(id=i.id, newsId=i.newsId, content=i.content) for i in self._repo.get_all() if i.newsId == news_id]

    def update_message(self, message_id: int, dto: MessageRequestTo) -> Optional[MessageResponseTo]:
        existing = self._repo.get_by_id(message_id)
        if not existing:
            return None
        self._ensure_news_exists(dto.newsId)
        updated_entity = Message(id=message_id, newsId=dto.newsId, content=dto.content)
        updated = self._repo.update(message_id, updated_entity)
        if not updated:
            return None
        return MessageResponseTo(id=updated.id, newsId=updated.newsId, content=updated.content)

    def delete_message(self, message_id: int) -> bool:
        return self._repo.delete(message_id)

File: 351004/Dashkevich/Security_lab/src/main/java/com/adashkevich/rest/lab/service/MessageService.java
package com.adashkevich.rest.lab.service;

import com.adashkevich.rest.lab.dto.request.MessageRequestTo;
import com.adashkevich.rest.lab.dto.response.MessageResponseTo;
import com.adashkevich.rest.lab.exception.NotFoundException;
import com.adashkevich.rest.lab.exception.ValidationException;
import com.adashkevich.rest.lab.mapper.MessageMapper;
import com.adashkevich.rest.lab.model.Message;
import com.adashkevich.rest.lab.repository.MessageRepository;
import com.adashkevich.rest.lab.repository.NewsRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class MessageService {
    private final MessageRepository repo;
    private final NewsRepository newsRepo;
    private final MessageMapper mapper;

    public MessageService(MessageRepository repo, NewsRepository newsRepo, MessageMapper mapper) {
        this.repo = repo;
        this.newsRepo = newsRepo;
        this.mapper = mapper;
    }

    public MessageResponseTo create(MessageRequestTo dto) {
        ensureNewsExists(dto.newsId);
        Message entity = mapper.toEntity(dto);
        Message saved = repo.save(entity);
        return mapper.toResponse(saved);
    }

    public List<MessageResponseTo> getAll() {
        return repo.findAll().stream().map(mapper::toResponse).toList();
    }

    public MessageResponseTo getById(Long id) {
        Message m = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Message not found", "40440"));
        return mapper.toResponse(m);
    }

    public MessageResponseTo update(Long id, MessageRequestTo dto) {
        Message existing = repo.findById(id)
                .orElseThrow(() -> new NotFoundException("Message not found", "40440"));

        ensureNewsExists(dto.newsId);
        existing.setNewsId(dto.newsId);
        existing.setContent(dto.content);

        Message updated = repo.update(id, existing);
        return mapper.toResponse(updated);
    }

    public void delete(Long id) {
        if (!repo.existsById(id)) {
            throw new NotFoundException("Message not found", "40440");
        }
        repo.deleteById(id);
    }

    private void ensureNewsExists(Long newsId) {
        if (newsId == null || !newsRepo.existsById(newsId)) {
            throw new ValidationException("newsId does not exist", "40020");
        }
    }
}
